src/main/java/org/mule/tools/maven/plugin/module/analyze/ModuleDiscoverer.java

```diff
@@ -0,0 +1,147 @@
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+
+import org.apache.maven.model.Resource;
+import org.apache.maven.project.MavenProject;
+
+/**
+ * Discovers the Mule modules used on the Maven project under analysis
+ *
+ * @since 1.0
+ */
+public class ModuleDiscoverer {
+
+  private final ModuleFactory moduleFactory = new ModuleFactory();
+
+  /**
+   * Discovers the module corresponding to the Maven project under analysis
+   *
+   * @param project project being analyzed.
+   * @param analyzerLogger collects all the logging information generated during the project analysis
+   * @return a module corresponding to the project being analyzed, null if the project is not a Mule module
+   * @throws ModuleApiAnalyzerException
+   */
+  public Module discoverProjectModule(MavenProject project, AnalyzerLogger analyzerLogger)
+      throws ModuleApiAnalyzerException {
+    Module module = null;
+    Properties properties = getModuleProperties(project);
+    if (properties != null) {
+      try {
+        module = moduleFactory.create(analyzerLogger, (String) properties.get("module.name"), properties);
+      } catch (IOException e) {
+        throw new ModuleApiAnalyzerException("Cannot read project's mule-module.properties", e);
+      }
+    }
+
+    return module;
+  }
+
+  /**
+   * Discovers all the Mule modules used as dependencies on the Maven project under analysis
+   *
+   * @param project project being analyzed.
+   * @param analyzerLogger collects all the logging information generated during the project analysis
+   * @param projectModuleName name of the module that corresponds to the project being analyzed
+   * @return a list containing all the Mule modules that are dependencies of the analyzed project.
+   * @throws ModuleApiAnalyzerException
+   */
+  public List<Module> discoverExternalModules(MavenProject project, AnalyzerLogger analyzerLogger,
+                                              String projectModuleName)
+      throws ModuleApiAnalyzerException {
+    final List<Module> result = new LinkedList<>();
+
+    Set<URL> urls = new HashSet<>();
+    List<String> elements;
+    try {
+      elements = project.getRuntimeClasspathElements();
+      elements.addAll(project.getCompileClasspathElements());
+
+      for (String element : elements) {
+        urls.add(new File(element).toURI().toURL());
+      }
+
+      ClassLoader contextClassLoader = URLClassLoader.newInstance(
+                                                                  urls.toArray(new URL[0]),
+                                                                  Thread.currentThread().getContextClassLoader());
+
+      try {
+        final Enumeration<URL> resources = contextClassLoader.getResources("META-INF/mule-module.properties");
+        while (resources.hasMoreElements()) {
+          final URL url = resources.nextElement();
+          Properties properties = loadProperties(url);
```

Comment:
Classloader returns URLs and I will need to open a connection to load the file, so it must use URLs anyway